src/context/node.js

```diff
@@ -0,0 +1,414 @@
+   * @return {?ContextNode}
+   */
+  static closest(node, excludeSelf = false) {
+    let n = node;
+    while (n) {
+      // Check if a node is a candidate to be returned.
+      if (n != node || !excludeSelf) {
+        if (n[NODE_PROP]) {
+          // Already a discovered node.
+          return /** @type {!ContextNode} */ (n[NODE_PROP]);
+        }
+        const {nodeType} = n;
+        if (nodeType == DOCUMENT_NODE || nodeType == FRAGMENT_NODE) {
+          // A context node is always created for a root. Due to this, a
+          // non-root element is always at least attached to a root. This
+          // allows for quick discovery and reattachment when new information
+          // becomes available.
+          return ContextNode.get(n);
+        }
+        if (nodeType == ELEMENT_NODE && startsWith(n.tagName, AMP_PREFIX)) {
+          // An AMP node will always have a context node backing it at some
+          // point.
+          return ContextNode.get(n);
+        }
+      }
+      // Navigate up the DOM tree. Notice that we do not automatically go over
+      // a root node boundary.
+      const assignedSlot =
+        /** @type {?Node|undefined} */ (n[ASSIGNED_SLOT_PROP]) ||
+        n.assignedSlot;
+      if (assignedSlot) {
+        n = assignedSlot;
+      } else {
+        n = n.parentNode;
+      }
+    }
+    return null;
+  }
+
+  /**
+   * Direct slot assignment. Works the same way as shadow slots, but does not
+   * require a shadow root. Automatically starts the discovery phase for the
+   * affected nodes.
+   *
+   * See `Element.assignedSlot` API.
+   *
+   * @param {!Node} node The target node.
+   * @param {!Node} slot The slot to which the target node is assigned.
+   */
+  static assignSlot(node, slot) {
+    if (node[ASSIGNED_SLOT_PROP] == slot) {
+      return;
+    }
+    node[ASSIGNED_SLOT_PROP] = devAssert(slot);
+    forEachContained(node, (cn) => cn.discover());
+  }
+
+  /**
+   * Unassigns the direct slot previously done by the `assignSlot` call.
+   * Automatically starts the discovery phase for the affected nodes.
+   *
+   * @param {!Node} node The target node.
+   * @param {!Node} slot The slot from which the target node is assigned.
+   */
+  static unassignSlot(node, slot) {
+    if (node[ASSIGNED_SLOT_PROP] != slot) {
+      return;
+    }
+    delete node[ASSIGNED_SLOT_PROP];
+    forEachContained(node, (cn) => cn.discover());
+  }
+
+  /**
+   * Creates the context node and automatically starts the discovery process.
+   *
+   * @param {*} privateOnly
+   * @param {!Node} node
+   */
+  constructor(privateOnly, node) {
+    devAssert(privateOnly === PRIVATE_ONLY);
+
+    /** @const @private {!Node} */
+    this.node_ = node;
+
+    /** @private {?ContextNode} */
+    this.parent_ = null;
+
+    /** @private {boolean} */
+    this.parentOverriden_ = false;
+
+    /** @private {boolean} */
+    this.isRoot_ = node.nodeType == DOCUMENT_NODE;
```

Comment:
Are `DocumentFragment`s not roots? I'm assuming no, they'll inherit from the `Document`. But then, are FIE's roots? I'd think not.

I think either everything has to be a root, or only the top-level `Document` can be a root.

Reply:
So, this is slightly more manual. Only a document automatically becomes a root b/c that's the only way we can do everything fully automatically. The fragments (and any other disconnected node) would simply be a disconnected tree until either `setParent` or `setIsRoot` API is called. The `setParent` will connect a fragment to an existing tree (e.g. a Bento component can connect shadow root this way). The `setIsRoot` API will make a fragment a functional root (e.g. a shadow doc can be started this way).

I believe this is a good compromise between being completely magically automatic and not crossing the boundaries unless "allowed". Additionally, a disconnected subtree (w/o root) does almost zero work. This reduces the CPU use until the subtree is connected to an actual root and avoids weird race conditions when a fragment is trying to use itself until the creator is ready to start it.

The best analogy to think about this is a orphaned DOM element or fragment. A disconnected subtree like this is "mostly" inert (though it's of course not always true) until it's actually added to the active document, in which cases it starts to automatically participate in layout, style calc, etc. Thus, before a subtree is added to the active document, `getComputedStyle()` returns mostly nothing. After it's added to the active document - `getComputedStyle()` starts returning actual values.